File: ch5/clip1/boj1917/Main.java
import java.util.Scanner;
import java.util.Stack;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = 6;
        for (int tc = 1; tc <= 3; tc++) {
            int[][] board = new int[n + 1][n + 1];
            int[][] check = new int[n + 1][n + 1];
            int[] diceCheck = new int[n + 1]; // 좌표 체크와 주사위 숫자를 체크할 배열 생성

            int[] dr = {-1, 0, 1, 0};
            int[] dc = {0, 1, 0, -1};
            final int UP = 0;
            final int RIGHT = 1;
            final int DOWN = 2;
            final int LEFT = 3;

            int startR = 0, startC = 0;
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    board[i][j] = sc.nextInt();
                    if (board[i][j] == 1) {
                        startR = i;
                        startC = j;
                    }
                }
            }

            Dice dice = new Dice();
            Stack<Pair<Point, Dice>> stack = new Stack<>();
            stack.push(new Pair<>(new Point(startR, startC), dice));
            diceCheck[dice.getTop()] = 1;
            check[startR][startC] = 1;
            while (!stack.isEmpty()) {
                Pair<Point, Dice> pair = stack.pop();
                Point now = pair.first;
                Dice nowDice = pair.second;
                diceCheck[nowDice.getTop()] = 1;
                for (int i = 0; i < 4; i++) {
                    Point next = new Point(now.r + dr[i], now.c + dc[i]);
                    if (isRange(next.r, next.c, n) && board[next.r][next.c] != 0 && check[next.r][next.c] == 0) {
                        Dice nextDice = new Dice(nowDice.getCube());
                        switch (i) {
                            case UP -> nextDice.up();
                            case RIGHT -> nextDice.right();
                            case DOWN -> nextDice.down();
                            case LEFT -> nextDice.left();
                        }
                        check[next.r][next.c] = 1;
                        stack.push(new Pair<>(new Point(next.r, next.c), nextDice));
                    }
                }
            }
            boolean isCube = true;
            for(int i = 1; i <= 6; i++) {
                if(diceCheck[i] == 0) {
                    isCube = false;
                    break;
                }
            }
            System.out.println(isCube ? "yes" : "no");

        }
    }

    static boolean isRange(int r, int c, int n) {
        return r >= 1 && r <= n && c >= 1 && c <= n;
    }
}
class Pair<T, U> {
    T first;
    U second;

    public Pair(T first, U second) {
        this.first = first;
        this.second = second;
    }
}

class Point {
    int r;
    int c;

    public Point(int r, int c) {
        this.r = r;
        this.c = c;
    }
}

class Dice {
    private int[] cube;
    /*
        _____
       / 1 / |
       ----|3|
       | 2 |/
       ----/
    */
    public Dice() {
        cube = new int[7];
        for (int i = 1; i <= 6; i++) {
            cube[i] = i;
        }
    }

    public Dice(int[] c) {
        cube = new int[7];
        for (int i = 1; i <= 6; i++) {
            cube[i] = c[i];
        }
    }

    public int getTop() {
        return cube[1];
    }

    public int[] getCube() {
        return cube;
    }

    public void up() {
        int temp = cube[1];
        cube[1] = cube[2];
        cube[2] = cube[6];
        cube[6] = cube[5];
        cube[5] = temp;
    }

    public void down() {
        int temp = cube[1];
        cube[1] = cube[5];
        cube[5] = cube[6];
        cube[6] = cube[2];
        cube[2] = temp;
    }

    public void left() {
        int temp = cube[1];
        cube[1] = cube[3];
        cube[3] = cube[6];
        cube[6] = cube[4];
        cube[4] = temp;
    }

    public void right() {
        int temp = cube[1];
        cube[1] = cube[4];
        cube[4] = cube[6];
        cube[6] = cube[3];
        cube[3] = temp;
    }
}

File: ch5/clip1/boj1917/main.py
import sys

def is_range(r, c, n):
    return 1 <= r <= n and 1 <= c <= n

def solve(board):
    n = 6
    dice_check = [0] * (n + 1)

    dr = [-1, 0, 1, 0]
    dc = [0, 1, 0, -1]

    UP, RIGHT, DOWN, LEFT = 0, 1, 2, 3

    def up(c):
        temp = c[0]
        c[0] = c[1]
        c[1] = c[5]
        c[5] = c[4]
        c[4] = temp

    def down(c):
        temp = c[0]
        c[0] = c[4]
        c[4] = c[5]
        c[5] = c[1]
        c[1] = temp

    def left(c):
        temp = c[0]
        c[0] = c[2]
        c[2] = c[5]
        c[5] = c[3]
        c[3] = temp

    def right(c):
        temp = c[0]
        c[0] = c[3]
        c[3] = c[5]
        c[5] = c[2]
        c[2] = temp

    board = [[0] * (n + 1) for _ in range(n + 1)]
    check = [[0] * (n + 1) for _ in range(n + 1)]
    dice_check = [0] * (n + 1)

    start_r, start_c = 0, 0
    for i in range(1, n + 1):
        row = list(map(int, sys.stdin.readline().split()))
        for j in range(1, n + 1):
            board[i][j] = row[j - 1]
            if board[i][j] == 1:
                start_r = i
                start_c = j

    stack = [(start_r, start_c, [1, 2, 3, 4, 5, 6])]
    dice_check[1] = 1
    check[start_r][start_c] = 1

    while stack:
        now_r, now_c, now_dice = stack.pop()
        dice_check[now_dice[0]] = 1

        for i in range(4):
            next_r = now_r + dr[i]
            next_c = now_c + dc[i]

            if is_range(next_r, next_c, n) and board[next_r][next_c] != 0 and check[next_r][next_c] == 0:
                next_dice = now_dice[:]
                if i == UP:
                    up(next_dice)
                elif i == RIGHT:
                    right(next_dice)
                elif i == DOWN:
                    down(next_dice)
                elif i == LEFT:
                    left(next_dice)

                check[next_r][next_c] = 1
                stack.append((next_r, next_c, next_dice))

    is_cube = all(dice_check[1:])
    print("yes" if is_cube else "no")

for _ in range(3):
    solve([])
